src/Xamarin.Android.Build.Tasks/Utilities/MarshalMethodCecilAdapter.cs

```diff
@@ -0,0 +1,253 @@
+#nullable enable
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.Android.Build.Tasks;
+using Microsoft.Build.Utilities;
+using Mono.Cecil;
+using Xamarin.Android.Tools;
+
+namespace Xamarin.Android.Tasks;
+
+class MarshalMethodCecilAdapter
+{
+	public static NativeCodeGenStateCollection? GetNativeCodeGenStateCollection (TaskLoggingHelper log, ConcurrentDictionary<AndroidTargetArch, NativeCodeGenState>? nativeCodeGenStates)
+	{
+		if (nativeCodeGenStates is null) {
+			log.LogDebugMessage ($"No {nameof (NativeCodeGenState)} found");
+			return null;
+		}
+
+		var collection = new NativeCodeGenStateCollection ();
+
+		// Convert each architecture
+		foreach (var kvp in nativeCodeGenStates) {
+			var arch = kvp.Key;
+			var state = kvp.Value;
+			var obj = CreateNativeCodeGenState (arch, state);
+			collection.States.Add (arch, obj);
+		}
+
+		return collection;
+	}
+
+	static NativeCodeGenStateObject CreateNativeCodeGenState (AndroidTargetArch arch, NativeCodeGenState state)
+	{
+		var obj = new NativeCodeGenStateObject ();
+
+		if (state.Classifier is null)
+			return obj;
+
+		foreach (var group in state.Classifier.MarshalMethods) {
+			var methods = new List<MarshalMethodEntryObject> ();
```

Comment:
I guess some of the `List<t>` creation, we would know the capacity:
```suggestion
			var methods = new List<MarshalMethodEntryObject> (state.Classifer.MarshalMethods.Count);
```